Refactoring display of slides in CaseViewer and loading of instances and new UI controls for multiSamples blending

Punzo opened the pull request:
@hackermd  Implement https://github.com/MGHComputationalPathology/slim/issues/22, doing the following assumptions:
https://github.com/MGHComputationalPathology/slim/pull/25/files#diff-25b188602cbdb7a70c7b8d7736bdbd2bab7e94a4e591353ca29b05b4529a67b4R4-R13

Tested with https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/54

![image](https://user-images.githubusercontent.com/7985338/120222940-10799b80-c241-11eb-848b-fae9a7c90e96.png)
![image](https://user-images.githubusercontent.com/7985338/120223040-3acb5900-c241-11eb-8ca0-887f5e712d9e.png)


I will continue on https://github.com/MGHComputationalPathology/slim/issues/8 on this draft PR.

hackermd commented on src/utils/fromSeriesListToSlideList.tsx:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  if (volumeMetadata.length > 0) {
+    const slideState: dmv.metadata.SlideState = {
```
Based on our discussion on #22, I think we should refer to this group as an "acquisition".

hackermd commented on src/components/ChannelItem.tsx:
```diff
@@ -0,0 +1,102 @@
+
+
+interface ChannelItemProps {
+  identifier: string,
+  description: string,
```
We may want to describe the "channel" in further detail using information from the Specimen Preparation Sequence (in particular the stain that was used).

@dclunie I assume the items of Specimen Preparation Sequence are structured according to the current version of [TID 8001 Specimen Preparation](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_C.html#sect_TID_8001) and we would need to look for the "Using substance" content item of [TID 8003 Specimen Staining](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_C.html#sect_TID_8003). Is that assumption correct?

Punzo replied:
done

hackermd reviewed (changes requested):
This already looks really nice! Great progress @Punzo!

I did a first round of review without delving into all of the implementation details. At this point, I would like to provide some high level comments related to the overall design and terminology:

1. I think we should use the term "acquisition" rather than "slide" (see https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-852467997 for background).
2. I may be missing something, but I think we could perform the grouping of VL Whole Slide Microscopy Image instances into acquisitions once in the `CaseViewer` component that then pass `acquisition` objects (type is currently called `slideState`) to the other components (the `SlideViewer` and its children).

Punzo commented on public/config/local.js:
```diff
@@ -4,7 +4,7 @@ window.config = {
     {
       id: "local",
-      url: "http://localhost:8008/dicomweb",
+      url: "http://34.68.90.36/",
```
thanks for catching this, commited by error!

hackermd replied:
Could you please change it back to localhost?

Punzo replied:
sorry I always forget to change them back after testing, done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/d39920423617f81050bfaa213a7e35226ea46931

Punzo commented on types/dicom-microscopy-viewer/index.d.ts:
```diff
@@ -267,13 +281,31 @@ declare module 'dicom-microscopy-viewer' {
       constructor ({ metadata: Metadata }: object)
     }
 
+    export interface SeriesState {
+      Series: Series
+      VolumeMetadata: object[]
+      LabelMetadata: object[]
+      OverviewMetadata: object[]
+    }
+
+    export interface SlideState {
+      Key: string
+      IsMultiChannel: boolean
+      MultiChannelsSeriesUIDs: string[]
+      Description: string
+      VolumeMetadata: object[]
+      LabelMetadata: object[]
+      OverviewMetadata: object[]
+    }
+
```
done

Punzo commented:
> This already looks really nice! Great progress @Punzo!
> 
> I did a first round of review without delving into all of the implementation details. At this point, I would like to provide some high level comments related to the overall design and terminology:
> 
> 1. I think we should use the term "acquisition" rather than "slide" (see [#22 (comment)](https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-852467997) for background).
> 2. I may be missing something, but I think we could perform the grouping of VL Whole Slide Microscopy Image instances into acquisitions once in the `CaseViewer` component that then pass `acquisition` objects (type is currently called `slideState`) to the other components (the `SlideViewer` and its children).

Thanks for the review! I have applied the comments in https://github.com/MGHComputationalPathology/slim/pull/25/commits/706fe4c61d0deeb06477119a32ae448a343399f5 and replied to https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-852467997, see https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-852813751

I will push soon in this same PR the UI upgrade for selecting the bleding parameters for Samples in the case of Multiplexed-Samples

Punzo commented:
@hackermd I have finished the UI part with commit https://github.com/MGHComputationalPathology/slim/pull/25/commits/36c31e079739d4e2539cf8ab2397cd97d83984f2

@pieper @fedorov  see a video here:
https://www.dropbox.com/s/ncpawj9yd9idkqt/2021-06-02%2015-25-29.mp4?dl=0

Still to do: finalizing the series/instance grouping with a better logic, i.e. comment https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-853008049
For this final part I need new or modified datasets since the series/image insatnces of the datasets server http://34.68.90.36/ studyInstanceUID = '1.3.6.1.4.1.5962.99.1.2103930081.1286074986.1595536829665.3.0' has different Frame of reference UID values

Punzo commented:
Added proper logic for grouping instances, @hackermd please review. I tested locally (with my local DCM4CHEE) with the new data from David (i.e. https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-854919111).

Punzo commented:
NOTE: at the moment we use `opticalPathIdentifier` to identify channels, in future we should switch to a `opticalPathUID` which will be unique in all cases, see https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-854642273.

P.S.: `opticalPathUID` does not currently exist in the DICOM standard, we would need to submit a correction.

hackermd commented on src/utils/fromSeriesListToSlideList.tsx:
```diff
@@ -0,0 +1,227 @@
+ * @params seriesList - array of series states
+ * @returns slideList, an array of slides.
+ */
+export const fromSeriesListToSlideList = (
```
For example, the handling of VOLUME, LABEL, and OVERVIEW metadata could probably be handled in separate functions.

hackermd commented on src/utils/fromSeriesListToSlideList.tsx:
```diff
@@ -0,0 +1,227 @@
+import * as dmv from 'dicom-microscopy-viewer'
+
+import {SeriesState, Slide} from './types'
```
I would suggest renaming this file to `slide.tsx` and define a `Slide` class. The `fromSeriesListToSlideList()` could then become the constructor of the `Slide` class. That would in my opinion result in a cleaner and more intuitive interface.

hackermd replied:
Furthermore, I would suggest moving `slide.tsx` into a separate directory called `data` as a sibling directory of `components`.

hackermd commented on src/utils/fromSeriesListToSlideList.tsx:
```diff
@@ -0,0 +1,227 @@
+        labelMetadata: [],
+        overviewMetadata: [],
+        seriesUIDs: [series.Series.SeriesInstanceUID],
+        keyOpticalPathIdentifier: '',
```
I am wondering whether we should already introduce the concept of an Optical Path UID here in the application and rename the property to `opticalPathUID`. Thoughts?

Punzo replied:
`keyOpticalPathIdentifier` is the Optical Path identifier that will be used to identify the first channel to be loaded and visualized when opening a study with "multi channels".

Note that if initiallySelectedSeriesInstanceUID is provided, than the keyOpticalPathIdentifier will be teh one of the first "channel" identified in the series.

No I don't think we should use `opticalPathUID`. This because at the moment we don't have any unique dicom attribute that identify a channel. OpticalPathIdentifier attribute is the best shot and we currently use it. Thefore using another  variable name will be just very confusing. Of course as soon as we have an opticalPathUID attribute (fully unique) in the standard we can switch to use that one.

hackermd replied:
We will need a `opticalPathUID` anyways. Naming the property `keyOpticalPathIdentifier` is more confusing in my opinion.

I would suggest using `opticalPathUID` internally within the app and in parallel working with @dclunie on including the attribute standard.

hackermd replied:
I think I misunderstood the intended use of `keyOpticalPathidentifier`. The term is a bit confusing. What do you think about calling it `selectedOpticalPathidentifier`? There are several existing properties in the app named `selected*` so that would be consistent.

In addition, I would suggest renaming `key` to `selectedSeriesUID` for consistency and improved clarity.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/9fe0a311e6aa9ed196267ddfd880536d5043a4e9

hackermd reviewed (changes requested):
Thanks @Punzo! It's coming along nicely. I only have a few suggestions for further improving the interface and the structure of the package.

I also noticed a couple of coding style issues and would appreciate if you could fix the linting errors reported by `yarn lint`.

hackermd commented:
@Punzo I just realized that there were more changes that were not displayed in the `Files changed` tab due to a filter. I will continue the review and may add more comments/suggestions.

hackermd commented on src/utils/types.tsx:
```diff
@@ -0,0 +1,22 @@
+import * as dmv from 'dicom-microscopy-viewer'
+
+export interface SeriesState {
+  Series: dmv.metadata.Series
```
Is this property still required? Most of the information should also be contained in `volumeMetadata`, `labelMetadata` and `overviewMetadata`.

hackermd replied:
I wonder whether `SeriesState` is required any longer at all? Now that we group image instances into `Slide` instances rather than series, we wouldn't need to pass around the series information, would we?

Punzo replied:
we still need them, because for the url redirection we use the series uid. How it works, it is that I store all the series uid in the slide (`seriesUIDs` variable in the slide interface). Then when the url is opened with a series uid, we look in the list of series uid in the slides and we show the slide that has the series uid selected in its seriesUIDs list.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/1244c8290b14be177b88c26c1f696b31bba54a5c

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -52,32 +56,75 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+
+    this.setState(state => ({
+      slideList: slideList,
+      seriesList: seriesList,
```
I wonder whether we need `seriesList`, since all the relevant metadata should also be contained in `slideList`.

Punzo replied:
we still need them only for the redirection of the url, which is based on series UIDs, see also https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r656126006. Of course if we change the redirection from `/studies/${this.props.studyInstanceUID}/series/${seriesInstanceUID}` to something else we could eliminate all the storing of the seriesList. However, I would leave as it is, because at the moment we don't have an unique UID to use to identify slides and it would be very confunsing for the users to use the Optical Path Identifier.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/1244c8290b14be177b88c26c1f696b31bba54a5c

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -52,32 +56,75 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+    return seriesList
   }
 
   handleSeriesSelection (
```
This should now probably be called `handleSlideSelection`.

Punzo replied:
I think is is better to still name it `handleSeriesSelection`, this because we don't have an unique UID to select slides. At teh momement we actually select series and then teh software choose to which slide is related the series (slide stores all the series UIDs) see also previous comments https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r656129187 and https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r656126006

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -90,10 +137,11 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+    const firstSeriesState = this.state.seriesList[0] as SeriesState
+    const studyMetadata = firstSeriesState.Series as dmv.metadata.Study
 
     /* If a series is encoded in the path, route the viewer to this series.
```
Hmm...this approach is now problematic, since there may be multiple series. We should consider updating the routing mechanism. The question is what concept (and UID) we could use. There is not really an identifier for the "slide". The closest is probably the Frame of Reference UID. Thoughts?

Punzo replied:
as I wrote in the previous comments, yes we still redirect using seriesUIDs which do not identify slides. Probably Frame of Reference UID is good enough, but I would wait to have the OpitcalPathUID in the DICOM standard as you suggested. That would be unique, straightforward and it would not create confusion.

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -52,32 +56,75 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+      selectedSeriesInstanceUID);
+
+    this.setState(state => ({
+      slideList: slideList,
```
I would prefer if we'd call this property just `slides` rather than `slideList`. I used the `List` suffix for certain React components to indicate that they represent HTML lists.

Punzo replied:
called them slideArray, since slides is the name of the new class that handle series information to slides.

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+import SamplesList from './SamplesList'
+
+interface SampleItemProps {
+  opticalPathSequence: dmv.metadata.OpticalPathDescription,
```
The use of the term `opticalPathSequence` for the property is very confusing, since it is actually only an individual item.

```suggestion
  opticalPathDescription: dmv.metadata.OpticalPathDescription,
```

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+  opticalPathSequence: dmv.metadata.OpticalPathDescription,
+  specimenDescriptionSequence: dmv.metadata.SpecimenDescription,
+  viewer: dmv.viewer.VolumeImageViewer,
+  itemRemoveHandler: SamplesList["handleItemRemoveSample"]
```
Is this valid Typescript? Either way, I would prefer type annotating the function signature explicitly.

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+  }
+
+  handleClippingChange (
+    value: number[],
```
```suggestion
    values: number[],
```

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+  }
+
+  handleLimitChange (
+    value: number[],
```
```suggestion
    values: number[],
```

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+  ): void {
+    const identifier = this.props.opticalPathSequence.OpticalPathIdentifier
+    const blendInfo = this.props.viewer.getBlendingInformation
+      (identifier) as dmv.viewer.BlendingInformation
```
`BlendingInformation` is not defined under the `viewer` namespace, but under the `metadata` namespace (see [dicom-microscopy-viewer](BlendingInformation). This should be correctly reflected here.

hackermd replied:
It should actually go under the `channel` namespace rather than the `metadata` namespace. I would suggest changing it at this point, where probably nobody else is using this functionality..

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+
+interface SampleItemProps {
+  opticalPathSequence: dmv.metadata.OpticalPathDescription,
+  specimenDescriptionSequence: dmv.metadata.SpecimenDescription,
```
See comment above:

```suggestion
  specimenDescription: dmv.metadata.SpecimenDescription,
```

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+      const description = specimenDescription.SpecimenShortDescription
+      if (description) {
+        attributes.push({
+          name: 'Description',
```
Better be explicit here, since it may otherwise not be clear what is being described:

```suggestion
          name: 'Specimen Description',
```

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+              console.debug(
+                `parse specimen preparation step "${processingType}"`
+              )
+            } else {
+              if (doesCodeMatch(name, 'SCT', '17636008')) {
+                attributes.push({
+                  name: 'Surgical collection',
+                  value: value.CodeMeaning
+                })
+              } else if (doesCodeMatch(name, 'SCT', '430864009')) {
+                attributes.push({
+                  name: 'Fixative',
+                  value: value.CodeMeaning
+                })
+              } else if (doesCodeMatch(name, 'SCT', '430863003')) {
+                attributes.push({
+                  name: 'Embedding medium',
+                  value: value.CodeMeaning
+                })
+              } else if (doesCodeMatch(name, 'SCT', '424361007')) {
+                attributes.push({
+                  name: 'Stain',
+                  value: value.CodeMeaning
+                })
+              }
+            }
+          } else if (item.ValueType === 'TEXT') {
+            item = item as dcmjs.sr.valueTypes.TextContentItem
+            if (doesCodeMatch(name, 'SCT', '424361007')) {
+              attributes.push({
+                name: 'Stain',
+                value: item.TextValue
+              })
+            }
+          } else {
+            console.debug(`specimen preparation step #${index} not rendered`)
+          }
+        })
+      }
+    )
```
This looks like it's copied from `SpecimenItem`. I would not repeat all the information about surgical collection, fixative, and embedding. That should be shared across samples.

I would suggest showing information of all preparation steps but "staining" in `SpecimenItem` and only show the information of "staining" in `SampleItem`. In case of a color image, I would continue to show information of all preparation steps in `SpecimenItem`.

Moving forward, we may want to compute the diff across samples, and then show all common preparation steps in `SpecimenItem` and the variable preparations steps for the corresponding `SampleItem`. However, I would wait until we have standardized the encoding of specimen preparation for multiplexed IF (cc @dclunie for awareness).

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,393 @@
+              </Popover>
+            </Space>
+        
+            <Button type="primary" danger icon={<CloseSquareFilled />} onClick={this.handleRemoveSample}>
```
Instead of using a separate button, could we use a [clickable Batch](https://ant.design/components/badge/#components-badge-demo-link) and an `x` icon?

Punzo replied:
unfortunately ant-badge do not have onClick events (you can just put clickable links into them).

Another option can be to use a [popconfirm](https://ant.design/components/popconfirm/) when clicking on the card descriptions, but I think that the simple X button is much more intuitive.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/e0b698a2a58d9bbbec07aded39fdf8be76586df1

hackermd commented on src/components/SamplesList.tsx:
```diff
@@ -0,0 +1,166 @@
+      (item: dmv.metadata.VLWholeSlideMicroscopyImage) => {
+        return (
+          <SampleItem
+            key={item.OpticalPathSequence[0].OpticalPathIdentifier}
```
This may get us into trouble. See comment on introducing an Optical Path UID property.

Punzo replied:
I agree with you, but unitl we don't have Optical Path UID property in teh standard the OpticalPathIdentifier is the best shot. This is alreaby been tracked in https://github.com/MGHComputationalPathology/slim/issues/31

hackermd replied:
I tend argue that we need to assign an Optical Path UID, independent of whether it will be included in the standard. For now, the UID may be assigned dynamically and not persisted. We cannot rely on Optical Path Identifier, since it's not guaranteed to be unique.

dclunie replied:
If FrameOfReferenceUID and OpticalPathIdentifier together produce a combination that is unique within a StudyInstanceUID, would that not suffice?

hackermd replied:
> If FrameOfReferenceUID and OpticalPathIdentifier together produce a combination that is unique within a StudyInstanceUID, would that not suffice?

@dclunie I don't think that the combination would result in a unique value for each "channel"/"sample" either.

Punzo replied:
also having two parameters will make the interface with the dicom-microscopy-viewer much more complicated. 

P.S.: in general dicom-microscopy-viewer is currently using as well the Optical Path Identifier to identify channels. 

> We cannot rely on Optical Path Identifier, since it's not guaranteed to be unique.

again I see your point, but in dicom-microscopy-viewer we are using the Optical Path Identifier ( and we have to use the same parameter to access channels both in SLIM and dicom-microscopy-viewer, otherwise teh two will be poiting to two different things). 

I would say that we know it is not unique and we are tracking it in https://github.com/MGHComputationalPathology/slim/issues/31 and as soon as we have a better solution we can update the code both in dicom-microscopy-viewer and SLIM.

hackermd replied:
I would like to address this sooner rather than later, since this will result in a backwards incompatible change that will break the API of the dicom-microscopy-viewer library.

hackermd replied:
Based on our conversation from this morning, we'll continue assuming that the Optical Path Identifier will be unique.

@Punzo no change will be required in the application or the dicom-microscopy-viewer library. I will resolve this comment and other related comments.

@dclunie it would be great if we could improve the language in the standard. Maybe we could add an additional note to clarify how the attribute is intended to be used in the context of multiplexed immunofluorescence imaging and to which extent values should be unique to enable particular use cases.

Punzo commented on src/utils/fromSeriesListToSlideList.tsx:
```diff
@@ -0,0 +1,227 @@
+      }
+  
+      if (slide.opticalPathIdentifiersList.length > 1) {
+        slide.description = 'Multiplexed-Samples'
```
We still a string in some cases. For example all the test data on multi channel have no label/overview instances. Therefore we need at least a string in description as `Multiplexed-Samples`. I suggest to simply put always this string and it would be simply `Multiplexed-Samples` for monochorme multi channels, `RGB Slide` for a RGB image and `Monochorme Slide` for a monochorme image

Punzo commented:
@hackermd I have replied/applied to your commets.

TO DO:
1) there is still a main point unresolved: rerouting. At the moment is done with `seriesUID` in the url. I think we should updated this logic later on when we will have the `OpticalPathUID` in the standards since at the moment we don't have any real unique UID to identify `Slides`. 

In the current implementation all the seriesUIDs that "contribute" to the Slide are stored in the slide interface and when rerouting it is picked the Slide that have the selected seriesUID in its seriesUIDs list. 


See comments: 
https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r656126006 https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r656129187 https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r656130849 https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r656132891

2) release new version of https://github.com/MGHComputationalPathology/dicom-microscopy-viewer and update in SLIM the package.json.

hackermd commented:
Thanks @Punzo.

> there is still a main point unresolved: rerouting. At the moment is done with seriesUID in the url. I think we should updated this logic later on when we will have the OpticalPathUID in the standards since at the moment we don't have any real unique UID to identify Slides

I don't think the Optical Path UID would help us here. The problem we are facing is due to the fact that we are blending images that are contained in different [Series](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/chapter_A.html#sect_A.1.2.3). Per definition, the collection of images that belong to a "slide" share the same [Frame of Reference](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/chapter_A.html#sect_A.1.2.5), so we could use the Frame of Reference UID for routing. However, the Frame of Reference is not a child of the [Study](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/chapter_A.html#sect_A.1.2.2) in the [IOD Entity-Relationship Model](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/chapter_A.html#figure_A.1-1) and other images that are not part of the collection (and should thus not be blended) may have the same Frame of Reference UID. Therefore, we can't use the Frame of Reference UID either.

In conclusion, I tend argue that we should stick to using the Series Instance UID for routing and route each "series" that is part of a "slide" (collection of images with the same Frame of Reference UID and Container Identifier) to the corresponding `SlideViewer` instance. Ideally, the viewer would then also display the images (channel/sample) that are contained in the specified series (specified via the Series Instance UID in the URL).

Punzo commented:
> In conclusion, I tend argue that we should stick to using the Series Instance UID for routing and route each "series" that is part of a "slide" (collection of images with the same Frame of Reference UID and Container Identifier) to the corresponding `SlideViewer` instance. Ideally, the viewer would then also display the images (channel/sample) that are contained in the specified series (specified via the Series Instance UID in the URL).

I agree, at the moment it is already managed that when selecting a series in the url, the viewr will display the right slide that contains the instances from that series. So I think we are good to go on this.

hackermd commented on src/components/SlideList.tsx:
```diff
@@ -22,37 +22,66 @@ interface SlideListState {
-      selectedSeriesInstanceUID: this.props.initiallySelectedSeriesInstanceUID
-    }
+  state = {
+    selectedSeriesInstanceUID: '',
```
I would use `undefined` instead of an empty string. An empty string is not a valid DICOM UID

Punzo replied:
that property is declared as `string`, therefore `undefined` is not an accepted value. We would need to declare it as `any`, but I would avoid using `any` declaration.

hackermd replied:
Please use `selectedSeriesInstanceUID?: string` to indicate that the value may be `undefined`. This is explicit and idiomatic TypeScript.

hackermd replied:
Somehow the logic doesn't seem to fit yet. The value of the `selectedSeriesInstanceUID` should ultimately neither be `undefined`. However, that is currently only guaranteed because the value is set to `""` (empty string) at some point. The TypeScript therefore doesn't complain, but it would probably still break the app if no series has been selected.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/c80070ff1342f92a08c9c8bd5806ced64c901de0

hackermd commented on src/components/SlideList.tsx:
```diff
@@ -22,37 +22,66 @@ interface SlideListState {
+        key={slide.key}
+        slide={slide}
+        client={this.props.client}
+      />
```
```suggestion
      const slideItem = (
        <SlideItem
          key={slide.key}
          slide={slide}
        client={this.props.client}
        />
      )
```

Punzo replied:
solved in https://github.com/MGHComputationalPathology/slim/pull/25/commits/89b2ea4a2905d6de7f99fdb36109e201674b4248

hackermd commented on src/components/SlideList.tsx:
```diff
@@ -22,37 +22,66 @@ interface SlideListState {
+        opticalPathIdentifiersList: [],
+        keyOpticalPathIdentifier: '',
+        description: ''
+      };
```
Why would we add an "empty" slide?

Punzo replied:
all this if condition is not needed, not sure why I have added it. I just removed it in https://github.com/MGHComputationalPathology/slim/pull/25/commits/97fa8eae7736b345d8206d8f836006923799dd59

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -37,13 +39,15 @@ interface ViewerProps extends RouteComponentProps {
 interface ViewerState {
-  series: dmv.metadata.Series[]
+  seriesArray: SeriesState[]
+  slideArray: Slide[]
```
```suggestion
  slides: Slide[]
```

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/13a6c2cc0f5a262ff9878e372771b33d6c5a6143

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,274 @@
+/**
+ * Arranges all metadata to create Slides  
+ */
+class Slides {
```
You currently use the `Slides` class as follows:

```javascript
const slides = new Slides(seriesArray, selectedSeriesInstanceUID)
const slideArray = slides.getSlideArray()
```

This could just become

```javascript
const slides = createSlides(seriesArray, selectedSeriesInstanceUID)
```

or something along those lines

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/13a6c2cc0f5a262ff9878e372771b33d6c5a6143

Punzo commented:
> I also noticed a couple of coding style issues and would appreciate if you could fix the linting errors reported by `yarn lint`.

done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/89b2ea4a2905d6de7f99fdb36109e201674b4248

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -52,32 +56,76 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+      const overviewMetadata: object[] = []
+      retrievedMetadata.forEach(item => {
+        const instance = dmv.metadata.formatMetadata(item) as dmv.metadata.Instance
+        if (instance.ImageType !== undefined && 
```
Why would `ImageType` be `undefined`? This is a type 1 attribute

Punzo replied:
It was defined in the metadata types as optional, fixed in https://github.com/MGHComputationalPathology/slim/pull/25/commits/b354a6cd397825c40cecfb13ad2d382b9334a36f

Punzo commented on src/components/CaseViewer.tsx:
```diff
@@ -103,7 +152,7 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
       selectedSeriesInstanceUID = fragments[4]
     } else {
-      const seriesMetadata = this.state.series[0] as dmv.metadata.Series
+      const seriesMetadata = firstSeriesState.Series as dmv.metadata.Series
```
this is already done, if in the url path is provided the series UID, then CaseViewer will display the slide correspondent to that series. the first series is used only if the series string is missing in the url (i.e. when the user simply open the study by clicking on the UI item of the study list at the beginning). I think this is teh right behaviour, i.e. here is simply taking the first series if it was not specified.

Punzo commented on src/components/CaseViewer.tsx:
```diff
@@ -37,13 +39,15 @@ interface ViewerProps extends RouteComponentProps {
 
 interface ViewerState {
-  series: dmv.metadata.Series[]
+  seriesArray: SeriesState[]
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/1244c8290b14be177b88c26c1f696b31bba54a5c

Punzo commented:
@hackermd I have applied the last review. Please have a look, thanks!

hackermd commented on src/components/SlideItem.tsx:
```diff
@@ -6,14 +6,14 @@ import { Menu } from 'antd'
 
 interface SlideItemState {
-  instances: dmv.metadata.Instance[]
   containerIdentifier: string
```
Why is `containerIdentifier` in `state`? This is not supposed to change. The information is available on `props.slide`. One should be able to just access it via `this.props.slide.overviewMetadata[0].ContainerIdentifier`.

```suggestion
```

hackermd replied:
If you want a separate property, then I would recommend either a [getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) or a private attribute.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/a51c9c79710a08163143cb6bc485c90f91bd998c

hackermd commented on src/components/SlideList.tsx:
```diff
@@ -22,37 +22,42 @@ interface SlideListState {
   }
 
   componentDidMount (): void {
+    this.setState(state => ({
+      selectedSeriesInstanceUID: this.props.initiallySelectedSeriesInstanceUID
+    }))
```
Why do you set `this.props.initiallySelectedSeriesInstanceUID` on the state in `componentDidMount` rather than the constructor? This is generally not a good idea (see [React docs](https://reactjs.org/docs/react-component.html#componentdidmount)).

> You may call setState() immediately in componentDidMount(). It will trigger an extra rendering, but it will happen before the browser updates the screen. This guarantees that even though the render() will be called twice in this case, the user won’t see the intermediate state. Use this pattern with caution because it often causes performance issues. In most cases, you should be able to assign the initial state in the constructor() instead.

Punzo replied:
done https://github.com/MGHComputationalPathology/slim/pull/25/commits/9122bd397bcd3bc3f8078c0a521695b65c249255

hackermd commented on src/components/SlideViewer.tsx:
```diff
@@ -120,6 +122,7 @@ interface SlideViewerProps extends RouteComponentProps {
 
 interface SlideViewerState {
+  activeSlide: Slide
   metadata: dmv.metadata.VLWholeSlideMicroscopyImage[]
```
That information is also available on `activeSlide`, isn't it? I would avoid having two separate state properties that both need to stay in sync. It is error prone. This may be a use case for a getter.

```suggestion
```

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/2b480cc755be204481548d3397dd1a94d2cdd585

hackermd commented on src/components/SlideViewer.tsx:
```diff
@@ -138,9 +141,25 @@ interface SlideViewerState {
   state = {
     isLoading: false,
     metadata: [],
+    activeSlide: this.slide,
```
Why do we need `this.slide` and `this.state.activeSlide`? I would suggest removing the `slide` attribute from the class and tracking the currently active slide via state.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/13e8f42278f4c3473105c364d904edc4e688dbbc

hackermd commented on src/components/SlideViewer.tsx:
```diff
@@ -138,9 +141,25 @@ interface SlideViewerState {
  * potentially one or more associated DICOM Series of DICOM SR documents.
  */
 class SlideViewer extends React.Component<SlideViewerProps, SlideViewerState> {
+  slide: Slide = {
+    key: '',
+    frameofReferenceUID: '',
+    containerIdentifier: '',
+    volumeMetadata: [],
+    labelMetadata: [],
+    overviewMetadata: [],
+    areImagesMonochrome: false,
+    isMultiplexedSamples: false,
+    seriesUIDsList: [],
+    opticalPathIdentifiersList: [],
+    keyOpticalPathIdentifier: '',
+    description: ''
+  }
+
```
I consider such objects with "nullable" values highly problematic. It would be much cleaner to just have an `undefined`. In this case, I don't think we actually need the attribute.

```suggestion
```

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/11f602ac3fcd6884fccb671852d8769b093b8d75

hackermd commented on src/components/SlideViewer.tsx:
```diff
@@ -316,75 +336,89 @@ class SlideViewer extends React.Component<SlideViewerProps, SlideViewerState> {
-      studyInstanceUID: this.props.studyInstanceUID,
-      seriesInstanceUID: this.props.seriesInstanceUID
-    }).then((retrievedMetadata): void => {
+    const slideArray = this.props.metadata
+
+    const slides = slideArray.filter(item => {
+      const slideItem = item
+      if (slideItem.seriesUIDsList.findIndex(uid => uid === this.props.seriesInstanceUID) !== -1) {
+        return true
+      }
+      return false
+    })
+
+    // at this point only 1 slide is selected
+    if (slides.length === 1) {
+      const slide = slides[0]
+
+      this.setState(state => ({
+        activeSlide: slide,
```
This should probably go into the constructor, i.e., the default `activeSlide` should be set upon construction.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/11f602ac3fcd6884fccb671852d8769b093b8d75

hackermd commented on src/components/SlideViewer.tsx:
```diff
@@ -316,75 +336,89 @@ class SlideViewer extends React.Component<SlideViewerProps, SlideViewerState> {
+      }))
+
       const series: dmv.metadata.VLWholeSlideMicroscopyImage[] = []
-      retrievedMetadata.forEach(item => {
+      slide.volumeMetadata.forEach(item => {
         const instance = dmv.metadata.formatMetadata(item) as dmv.metadata.VLWholeSlideMicroscopyImage
         series.push(instance)
       })
       this.setState((state) => ({ metadata: series }))
```
Could we handle this outside of this component such that `slide.volumeMetadata` is already formatted correctly?

hackermd replied:
See also comment on the need for storing `metadata` on the state object.

Punzo replied:
these lines are not necessary anumore, see https://github.com/MGHComputationalPathology/slim/commit/2b480cc755be204481548d3397dd1a94d2cdd585

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,261 @@
+ * @params labelMetadata - array of label metadata
+ * @params overviewMetadata - array of overview metadata
+ */
+class Slide {
+  frameofReferenceUID: string = ''
+  containerIdentifier: string = ''
+  areImagesMonochrome: boolean = false
+  isMultiplexedSamples: boolean = false
+  key: string = ''
+  seriesUIDsList: string[] = []
+  keyOpticalPathIdentifier: string = ''
+  opticalPathIdentifiersList: string[] = []
+  description: string = ''
+  volumeMetadata: object[] = []
+  labelMetadata: object[] = []
+  overviewMetadata: object[] = []
+}
```
There shouldn't be any default values. This will only get us into trouble. Whoever creates the `Slide` objects will be responsible for populating the required information.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/b457fc27cee8440c76bb315a09b78f4b05a20732

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,261 @@
+ */
+function createSlides (
+  instancesMetadataArray: InstancesMetadata[],
+  initiallySelectedSeriesInstanceUID: string = ''
```
Please use `undefined` instead of a `''`:

```suggestion
  initiallySelectedSeriesInstanceUID?: string
```

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/b457fc27cee8440c76bb315a09b78f4b05a20732

hackermd commented on src/utils/types.tsx:
```diff
@@ -0,0 +1,5 @@
+export interface InstancesMetadata {
```
Is that really something that we will need to re-use? If so, I would define it in `slide.tsx`.

Punzo replied:
it is necessary in passing the series array to slides trough the fetch functions. moved the declaration in slide.tsx https://github.com/MGHComputationalPathology/slim/pull/25/commits/b457fc27cee8440c76bb315a09b78f4b05a20732

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,261 @@
+ */
+function parseLabelMetadataFromListToSlide (
+  labelMetadataList: object[],
+  slide: Slide
```
I would suggest passing the relevant attributes directly

```suggestion
  frameOfReferenceUID: string,
  containerIdentifier: string
```

Same for the other functions.

Punzo replied:
done https://github.com/MGHComputationalPathology/slim/pull/25/commits/b457fc27cee8440c76bb315a09b78f4b05a20732

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,261 @@
+        isMultiplexedSamples: false,
+        description: ''
+      }
+      parseVolumeMetadataFromListToSlide(instancesMetadata.volumeMetadata, slide)
+      parseLabelMetadataFromListToSlide(instancesMetadata.labelMetadata, slide)
+      parseOverviewMetadataFromListToSlide(instancesMetadata.overviewMetadata, slide)
```
I am wondering whether these should be private methods on the `Slide` class. The constructor of the `Slide` class could accept the `instancesMetadata`, parse and extract the relevant information, and populate the attributes on the `Slide` object.

Punzo replied:
done https://github.com/MGHComputationalPathology/slim/pull/25/commits/b457fc27cee8440c76bb315a09b78f4b05a20732

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,261 @@
+      dmv.metadata.formatMetadata(metadata) as dmv.metadata.VLWholeSlideMicroscopyImage
+    if (j === 0) {
+      slide.frameofReferenceUID = instance.FrameOfReferenceUID
+    } else if (slide.frameofReferenceUID !== instance.FrameOfReferenceUID) {
+      console.warn('FrameOfReferenceUID of volume instance' +
+                   instance.SOPInstanceUID +
+                   ' does not correspond to slide FrameOfReferenceUID. ' +
+                   'The instance will be discarded.')
+      continue
+    }
+    if (j === 0) {
+      slide.containerIdentifier = instance.ContainerIdentifier
+    } else if (slide.containerIdentifier !== instance.ContainerIdentifier) {
+      console.warn('ContainerIdentifier of volume instance' +
+                   instance.SOPInstanceUID +
+                   ' does not correspond to slide ContainerIdentifier. ' +
+                   'The instance will be discarded.')
+      continue
+    }
```
This could probably be factored into a separate function/method which determines whether a given metadata object belongs to the slide.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/b457fc27cee8440c76bb315a09b78f4b05a20732

Punzo commented on src/components/SamplesList.tsx:
```diff
@@ -155,7 +155,10 @@ class SamplesList extends React.Component<SamplesListProps, SampleListState> {
-          <Select defaultValue='' style={{ width: 200 }} onChange={this.handleSelectChange} allowClear>
+          <Select
+            defaultValue='' style={{ width: 200 }} onChange={this.handleSelectChange}
+            value={this.state.selectedOpticalPathIdentifier} allowClear
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/e39270e2fb5421d964b9a4de310deb1546b9c9fe

Punzo commented on src/components/SamplesList.tsx:
```diff
@@ -23,7 +23,7 @@ interface SampleListState {
   state = {
     rerender: false,
-    opticalPathIdentifierToAdd: ''
+    selectedOpticalPathIdentifier: ''
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/e39270e2fb5421d964b9a4de310deb1546b9c9fe

Punzo commented on src/components/SamplesList.tsx:
```diff
@@ -0,0 +1,172 @@
+    this.props.viewer.activateOpticalPath(identifier)
+    this.props.viewer.showOpticalPath(identifier)
+    this.setState({
+      selectedOpticalPathIdentifier: ''
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/e39270e2fb5421d964b9a4de310deb1546b9c9fe

Punzo commented:
@hackermd thanks again for your last review, I applied the comments. We need only to address the non uniqueness of the Optical Path Identifier parameter.

hackermd commented on src/components/SamplesList.tsx:
```diff
@@ -0,0 +1,172 @@
+        if (item.OpticalPathSequence.length > 0) {
+          const index = opticalPaths.findIndex(
+            (property: dmv.metadata.VLWholeSlideMicroscopyImage) => {
+              return property.OpticalPathSequence[0].OpticalPathIdentifier ===
+                      item.OpticalPathSequence[0].OpticalPathIdentifier
```
We will continue to use Optical Path Identifier and assume that values are unique. Rather than adding a new Optical Path UID attribute to the standard, we will explore possibilities to further clarify uniqueness of expectations/requirements by receiving applications. See also [this comment](https://github.com/MGHComputationalPathology/slim/pull/25/files#r657055257).

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -52,32 +53,72 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+
+    const instancesMetadataArray = await this.fetchInstancesMetadataArray()
+
+    let selectedSeriesInstanceUID = ''
```
```suggestion
    let selectedSeriesInstanceUID = undefined
```

See [this comment](https://github.com/MGHComputationalPathology/slim/pull/25#discussion_r666149795)

hackermd replied:
This would allow the TypeScript compiler to catch the problem. Setting the value to an empty string will just propagate the problem.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/1fba907d84eddde77ccd787686a243b3af4061a2

hackermd reviewed (changes requested):
With the main issue regarding the Optical Path Identifier resolved, there are only two small minor comments remaining. The first is related to naming things and the second to type declarations and nullable values. Almost there :)

Punzo commented:
> With the main issue regarding the Optical Path Identifier resolved, there are only two small minor comments remaining. The first is related to naming things and the second to type declarations and nullable values. Almost there :)

Hi @hackermd I have applied your comments, thanks for the review! please have a look if we can merge.

Next steps To Do:
1) make a new release of https://github.com/MGHComputationalPathology/dicom-microscopy-viewer
2) update the package.json
3) make new release of slim

hackermd reviewed (commented):
Thanks @Punzo! Should I go ahead and perform final review and testing or are you still going to make further commits?

Punzo commented:
> Thanks @Punzo! Should I go ahead and perform final review and testing or are you still going to make further commits?

I have finished, please go ahead!

Punzo commented:
@igoroctaviano do you know why the Firebase deploy is skipped? I tried to create a new PR (https://github.com/MGHComputationalPathology/slim/pull/41) and I had the same issue.

hackermd commented on src/components/CaseViewer.tsx:
```diff
@@ -52,59 +53,109 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+     * Otherwise select the first series correspondent to the first slide contained in the study.
      */
     let selectedSeriesInstanceUID
     if (this.props.location.pathname.includes('series/')) {
       const fragments = this.props.location.pathname.split('/')
       selectedSeriesInstanceUID = fragments[4]
     } else {
```
This is no done twice in the same component. We should probably factor that out.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/6e2bf8f9ccabed9c59383f8657401f71cb150588

hackermd reviewed (changes requested):
Unfortunately, I ran into several issues/bugs when testing locally. See individual comments for details.

One of the reasons these bugs have remained undetected could be the liberal use of optional properties (i.e., properties whose values may be `undefined`). I would avoid that whenever possible, because it generally leads to undefined propagation, which can cause nasty bugs downstream that are difficult to detect and fix. Please revisit the type annotations of properties that you changed or introduced.

I further noticed that the `config/local.js` has changed. Please revert the local configuration to 

```javascript
window.config = {
    servers: [
    {
      id: "local",
      url: "http://localhost:8008/dicomweb",
      write: true
    }
  ],
  ...
}
```

Punzo commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,314 @@
+  ): void {
+    labelMetadataList.forEach((metadata) => {
+      if (this.instanceBelongToSlide(metadata)) {
+        this.overviewMetadata.push(metadata)
```
in https://github.com/MGHComputationalPathology/slim/pull/25/commits/384f3ead3b74de294b1a2064d5ca32ac878244dc I have changed the names of the methods and added few lines of descriptions in the docstring. 

I disagree in moving the `Slide` methods (`addInstanceMetadata`, `addLabelInstanceMetadata`, `addVolumeInstanceMetadata`, `addOverviewInstanceMetadata`) in separate functions. It would just overcomplicate the logic with no substancial advantage. I would say, it goes against a straightforward object oriented programming (since these methods/function needs to change several private attributes of the objects `Slide`). For example if you separate these methods from the Slide class in private functions, you would need then to pass the slide objects to the functions and then the functions would anyway to modify the 'state' of the object. Also to be noted is that `addLabelInstanceMetadata`, `addVolumeInstanceMetadata`, `addOverviewInstanceMetadata` are already private methods.

hackermd replied:
All these methods do is parse the metadata and then set values on the appropriate attributes on the `Slide` object. To me, this should either be the responsibility of the `Slide` constructor or the `createSlides()` factory function. I would prefer the latter as it would allow us to make `Slide` objects immutable (or least not provide methods for mutating an object after construction). The `Slide` objects determine a lot of downstream application behavior and I think we'll do us a big favor in making them immutable.

I would define a `SlideOptions` interface based on the attributes of `Slide` (`containerIdentifier`, `selectedSeriesUID`, etc.), create these option objects in `createSlides` using `_addInstanceMetadata()`, and then pass one object to the constructor of `Slide` for each "detected" slide. 

```typescript
interface SlideOptions {
  frameofReferenceUID: string
  containerIdentifier: string
  areImagesMonochrome: boolean
  isMultiplexedSamples: boolean
  selectedSeriesUID: string
  seriesUIDsList: string[] = []
  selectedOpticalPathidentifier: string
  opticalPathIdentifiersList: string[] = []
  description?: string
  volumeMetadata: VLWholeSlideMicroscopyImage[]
  labelMetadata: VLWholeSlideMicroscopyImage[] = []
  overviewMetadata: VLWholeSlideMicroscopyImage[] = []
}
```

None of the attributes required by the standard (`frameOfReferenceUID`, `containerIdentifier`, `volumeMetadata`, etc.) shall be optional, i.e, their values shall never be `undefined` or `[]`.

We can then add getter methods to the `Slide` class to make the properties read only.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/3a14700fadc8dd2ec993d897362a0af8e0770b06

Punzo commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,314 @@
+   *
+   * @params metadata - volume, label or overview instance
+   */
+  private instanceBelongToSlide (
```
done in https://github.com/MGHComputationalPathology/slim/commit/384f3ead3b74de294b1a2064d5ca32ac878244dc

Punzo commented on src/components/CaseViewer.tsx:
```diff
@@ -52,59 +53,109 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
 
   handleSeriesSelection (
-    { seriesInstanceUID }: { seriesInstanceUID: string }
+    { seriesInstanceUID }: { seriesInstanceUID?: string }
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/6e2bf8f9ccabed9c59383f8657401f71cb150588

Punzo commented on src/components/CaseViewer.tsx:
```diff
@@ -52,59 +53,109 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
-      `/studies/${this.props.studyInstanceUID}/series/${seriesInstanceUID}`
-    )
+    if (seriesInstanceUID === undefined || seriesInstanceUID === null) {
+      console.warn('switching series: seriesInstanceUID not found.')
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/6e2bf8f9ccabed9c59383f8657401f71cb150588

Punzo commented on src/components/SlideList.tsx:
```diff
@@ -1,73 +1,81 @@
+  metadata: Slide[]
   client: DicomWebManager
-  initiallySelectedSeriesInstanceUID: string
+  initiallySelectedSeriesInstanceUID?: string
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/6e2bf8f9ccabed9c59383f8657401f71cb150588

Punzo commented on src/components/CaseViewer.tsx:
```diff
@@ -52,59 +53,109 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
-      selectedSeriesInstanceUID = seriesMetadata.SeriesInstanceUID
+      const firstVolumeSeriesIstance =
+        dmv.metadata.formatMetadata(firstSlide.volumeMetadata[0]) as dmv.metadata.VLWholeSlideMicroscopyImage
+      selectedSeriesInstanceUID = firstVolumeSeriesIstance.SeriesInstanceUID
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/25f7e58df55d0a81da56b86661abbfb8d9264f96

Punzo commented on src/components/CaseViewer.tsx:
```diff
@@ -52,59 +53,109 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
     }
-    const studyMetadata = this.state.series[0] as dmv.metadata.Study
+    const studyMetadata =
+      dmv.metadata.formatMetadata(firstSlide.volumeMetadata[0]) as dmv.metadata.Study
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/25f7e58df55d0a81da56b86661abbfb8d9264f96

Punzo commented on src/components/CaseViewer.tsx:
```diff
@@ -52,59 +53,109 @@ class Viewer extends React.Component<ViewerProps, ViewerState> {
+    } else {
+      console.info(`switch to series "${seriesInstanceUID}"`)
+      this.props.history.push(
+        `/studies/${this.props.studyInstanceUID}/series/${seriesInstanceUID}`
```
fixed in https://github.com/MGHComputationalPathology/slim/pull/25/commits/9d6abac55191060390a6dc3f95e5d3af67b8fb5d. Please note also that this fix is needed too: https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/56

(tested with study 2.25.261030824872518877931236952174606351723  in https://idc-external-006.uc.r.appspot.com

Punzo commented:
> Unfortunately, I ran into several issues/bugs when testing locally. See individual comments for details.
> 
> One of the reasons these bugs have remained undetected could be the liberal use of optional properties (i.e., properties whose values may be `undefined`). I would avoid that whenever possible, because it generally leads to undefined propagation, which can cause nasty bugs downstream that are difficult to detect and fix. Please revisit the type annotations of properties that you changed or introduced.
> 
> I further noticed that the `config/local.js` has changed. Please revert the local configuration to
> 
> ```js
> window.config = {
>     servers: [
>     {
>       id: "local",
>       url: "http://localhost:8008/dicomweb",
>       write: true
>     }
>   ],
>   ...
> }
> ```

Hi @hackermd tahnsk for the review. I have fixed the main issue regarding the undefined variables. Please note that also this fix is needed https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/56. Regarding  `url: "http://localhost:8008/dicomweb" ` I just used the one in the current master which is set to `url: "https://idc-external-006.uc.r.appspot.com",`.

Punzo commented:
> Unfortunately, I ran into several issues/bugs when testing locally. See individual comments for details.
> 
> One of the reasons these bugs have remained undetected could be the liberal use of optional properties (i.e., properties whose values may be `undefined`). I would avoid that whenever possible, because it generally leads to undefined propagation, which can cause nasty bugs downstream that are difficult to detect and fix. Please revisit the type annotations of properties that you changed or introduced.
> 
> I further noticed that the `config/local.js` has changed. Please revert the local configuration to
> 
> ```js
> window.config = {
>     servers: [
>     {
>       id: "local",
>       url: "http://localhost:8008/dicomweb",
>       write: true
>     }
>   ],
>   ...
> }
> ```

Hi @hackermd thanks for the review.

I have fixed the main issue regarding the undefined variables. 

Please also note that this fix is needed as well: https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/56. 

Regarding the modification of `config/local.js`  (`url: "http://localhost:8008/dicomweb" ` into `url: "https://idc-external-006.uc.r.appspot.com",`), I just used the one in the current master which is set to `url: "https://idc-external-006.uc.r.appspot.com",`. Anywa, I switched back to the `url: "http://localhost:8008/dicomweb" `

hackermd commented:
> Regarding the modification of config/local.js (url: "http://localhost:8008/dicomweb" into url: "https://idc-external-006.uc.r.appspot.com",), I just used the one in the current master which is set to url: "https://idc-external-006.uc.r.appspot.com",. Anyway, I switched back to the url: "http://localhost:8008/dicomweb" 

Thanks! Sorry, it seems I missed this on another PR.

hackermd commented:
> Please also note that this fix is needed as well: MGHComputationalPathology/dicom-microscopy-viewer#56.

Thanks! I approved and merged this PR already. Will draft a new release once this PR is finalized.

hackermd commented:
> I have fixed the main issue regarding the undefined variables (nice catch!!! I previously tested only studies with a multichannels slide or 1 RGB slide).

Great! Let's briefly go through all optional parameters and double-check whether they should indeed be optional. For example, the `selectedSeriesInstanceUID` property on the `Slide` class is optional, but the application won't function properly if that property is not set. Same applies to `containerIdentifier`, which represents a type 1 DICOM attribute that is  guaranteed to be present in any VL Whole Sllide Microscopy Image instance. If the value cannot be determined upon construction of the `Slide` object, the application should just fail immediately and gracefully, rather than propagating the issue through the app.

Punzo commented:
> > I have fixed the main issue regarding the undefined variables (nice catch!!! I previously tested only studies with a multichannels slide or 1 RGB slide).
> 
> Great! Let's briefly go through all optional parameters and double-check whether they should indeed be optional. For example, the `selectedSeriesInstanceUID` property on the `Slide` class is optional, but the application won't function properly if that property is not set. Same applies to `containerIdentifier`, which represents a type 1 DICOM attribute that is guaranteed to be present in any VL Whole Sllide Microscopy Image instance. If the value cannot be determined upon construction of the `Slide` object, the application should just fail immediately and gracefully, rather than propagating the issue through the app.

Done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/3a14700fadc8dd2ec993d897362a0af8e0770b06

Punzo commented:
Commit https://github.com/MGHComputationalPathology/slim/pull/25/commits/25667f8eead3b4bc2bf2cee630a5f9e1fd65cb0b implement first step of https://github.com/MGHComputationalPathology/slim/issues/38

Punzo commented:
This PR has been tested both with OpenLayer versions 6.5.0 and 6.6.1. See also https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/57

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,471 @@
+  slideOptionsItem: SlideOptions,
+  initiallySelectedSeriesInstanceUID?: string
+): void {
+  const volumeInstanceReference =
+    _addVolumeInstanceMetadata(slideOptionsItem, instancesMetadata.volumeMetadata)
+  _addLabelInstanceMetadata(slideOptionsItem, instancesMetadata.labelMetadata)
+  _addOverviewInstanceMetadata(slideOptionsItem, instancesMetadata.overviewMetadata)
```
The logic of these methods is really hard to understand and error prone in my opinion.

I don't like the fact that the `SlideOptions` object is first created with invalid default values (e.g., empty strings for identifiers) and then passed around for its properties to be populated by different methods.

I would generally avoid functions/methods with side effects whenever possible and I would instead suggest first collecting the relevant information via dedicated methods and then creating the `SlideObject` object.

The `_addLabelInstanceMetadata()` and `_addOverviewInstanceMetadata()` methods basically just filter the metadata. This could probably be directly handled via [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) using `_doesInstanceBelongToSlide()`.

The `addVolumeInstanceMetadata()` method currently does more than just filtering, but I feel that this could be handled separately, i.e., the metadata could be filtered first and the other attributes could be determined afterwards.

I feel this approach would be much more intuitive and avoid all of the `=== undefined` and `=== ''` checks.

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/556a73606532c161180c101a812f644c6f809723

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,471 @@
+ */
+function _addVolumeInstanceMetadata (
+  slideOptionsItem: SlideOptions,
+  volumeMetadataList: object[]
```
In general, I would avoid using terms such as `list` and `array` in variable/property/key names. The type annotation already clearly indicates whether the value represents a single item or an array of items.

Punzo replied:
fixed in https://github.com/MGHComputationalPathology/slim/pull/25/commits/e3e5948bcfffe2ac2e784c8212f07e3083e571c4

hackermd commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,471 @@
+function _addVolumeInstanceMetadata (
+  slideOptionsItem: SlideOptions,
+  volumeMetadataList: object[]
+): dmv.metadata.VLWholeSlideMicroscopyImage | undefined {
```
Do we allow `Slide` objects without any VOLUME image instances? If not, I would also avoid `undefined` here. This should no longer be necessary after the suggested refactoring (see above comment).

Punzo replied:
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/556a73606532c161180c101a812f644c6f809723

hackermd reviewed (changes requested):
Thanks for the changes @Punzo. Already looks much cleaner. There are still a couple of things I suggest changing regarding the implementation of `createSlides()`. This method has become a critical piece that determines a lot of the downstream application logic. Therefore,  this part needs to be rock solid.

hackermd commented on src/components/SamplesList.tsx:
```diff
@@ -0,0 +1,178 @@
+          </Select>
+          <Button type='primary' icon={<AppstoreAddOutlined />} onClick={this.handleAddSample} />
+        </Space>
+        <h4 />
```
That is causing linting issues.

```suggestion
```

Punzo replied:
fixed in https://github.com/MGHComputationalPathology/slim/pull/25/commits/47c955fe1c0fac571edc07c57bc835473d236275

hackermd commented on src/components/SampleItem.tsx:
```diff
@@ -0,0 +1,369 @@
+      </div>
+    )
+
+    return (
+      <Space align='start'>
+        <div style={{ paddingLeft: '14px', paddingTop: '10px' }}>
+          <Space direction='vertical' align='end' size={100}>
+            <Space direction='vertical' align='end'>
+              <Switch
+                size='small'
+                checked={this.state.visible}
+                onChange={this.handleVisibilityChange}
+                checkedChildren={<FaEye />}
+                unCheckedChildren={<FaEyeSlash />}
+              />
+
+              <Popover placement='left' content={content} title='Blending Parameters'>
+                <Button type='primary' shape='circle' icon={<SettingOutlined />} />
+              </Popover>
+            </Space>
+          </Space>
+        </div>
+        <Space direction='horizontal' align='start'>
+          <Description
+            header={'ID: ' + identifier}
+            attributes={attributes}
+            selectable
+            hasLongValues
+          />
+          <a onClick={this.handleRemoveSample}>
+            <Badge>
+              <CloseCircleOutlined style={{ color: '#FF0000' }} />
+            </Badge>
+          </a>
+        </Space>
+      </Space>
+    )
```
```suggestion
    const removeSampleButton = (
      <CloseCircleOutlined
          style={{ color: '#FF0000' }}
          onClick={this.handleRemoveSample}
      />
    )

    return (
      <Space align='start'>
        <div style={{ paddingLeft: '14px', paddingTop: '10px' }}>
          <Space direction='vertical' align='end' size={100}>
            <Space direction='vertical' align='end'>
              <Switch
                size='small'
                checked={this.state.visible}
                onChange={this.handleVisibilityChange}
                checkedChildren={<FaEye />}
                unCheckedChildren={<FaEyeSlash />}
              />

              <Popover placement='left' content={content} title='Blending Parameters'>
                <Button type='primary' shape='circle' icon={<SettingOutlined />} />
              </Popover>
            </Space>
          </Space>
        </div>
        <Space direction='horizontal' align='start'>
          <Badge count={removeSampleButton} title='Remove sample'>
            <Description
              header={'ID: ' + identifier}
              attributes={attributes}
              selectable
              hasLongValues
            />
          </Badge>
        </Space>
      </Space>
    )
```

Punzo replied:
fixed in https://github.com/MGHComputationalPathology/slim/pull/25/commits/47c955fe1c0fac571edc07c57bc835473d236275

Punzo commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,471 @@
+      this.slideOptions.volumeMetadata[0]
+    ) as dmv.metadata.VLWholeSlideMicroscopyImage
+  }
+}
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/d0c0cb45e4197f19a031b1eb5fd19c5b56173f09

Punzo commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,471 @@
+): boolean {
+  const instance = dmv.metadata.formatMetadata(
+    metadata
+  ) as dmv.metadata.VLWholeSlideMicroscopyImage
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/2356fa22fd5253f4d394d2d804f12127170c224c

Punzo commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,471 @@
+
+    const instance = dmv.metadata.formatMetadata(
+      metadata
+    ) as dmv.metadata.VLWholeSlideMicroscopyImage
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/2356fa22fd5253f4d394d2d804f12127170c224c

Punzo commented on src/data/slides.tsx:
```diff
@@ -0,0 +1,471 @@
+    if (slideOptionsItem.selectedOpticalPathidentifier === '') {
+      slideOptionsItem.areImagesMonochrome = instanceIsMonochorme
+      slideOptionsItem.selectedOpticalPathidentifier = instanceOpticalPathIdentifier
+    } else if (instanceIsMonochorme !== slideOptionsItem.areImagesMonochrome) {
```
done in https://github.com/MGHComputationalPathology/slim/pull/25/commits/1d864a753f78c5093c37e15c4c3cca57fe75f0b3

Punzo commented on src/components/SamplesList.tsx:
```diff
@@ -0,0 +1,178 @@
+    this.props.viewer.showOpticalPath(identifier)
+    this.setState({
+      selectedOpticalPathIdentifier: undefined
+    })
```
https://github.com/MGHComputationalPathology/slim/pull/25/commits/47c955fe1c0fac571edc07c57bc835473d236275

Punzo commented:
> Thanks for the changes @Punzo. Already looks much cleaner. There are still a couple of things I suggest changing regarding the implementation of `createSlides()`. This method has become a critical piece that determines a lot of the downstream application logic. Therefore, this part needs to be rock solid.

Hi @hackermd, thansk for the review, I have applied the comments!

hackermd commented:
Thank you @Punzo. I have merged the PR into a separate `dev` branch, where I will perform additional tests before finally merging everything into `master`.
